xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,296 @@
+  @SuppressWarnings("unused")
+  @Nullable private XdsClient xdsClient;
+  private final int port;
+  private final ScheduledExecutorService timeService;
+
+  /**
+   * Factory method for creating a {@link XdsClientWrapperForServerSds}.
+   *
+   * @param port server's port for which listener config is needed.
+   * @param bootstrapper {@link Bootstrapper} instance to load bootstrap config.
+   * @param syncContext {@link SynchronizationContext} needed by {@link XdsClient}.
+   */
+  public static XdsClientWrapperForServerSds newInstance(
+      int port, Bootstrapper bootstrapper, SynchronizationContext syncContext) throws IOException {
+    Bootstrapper.BootstrapInfo bootstrapInfo = bootstrapper.readBootstrap();
+    final List<Bootstrapper.ServerInfo> serverList = bootstrapInfo.getServers();
+    if (serverList.isEmpty()) {
+      throw new NoSuchElementException("No management server provided by bootstrap");
+    }
+    final Node node = bootstrapInfo.getNode();
+    ScheduledExecutorService timeService = SharedResourceHolder.get(timeServiceResource);
+    XdsClientImpl xdsClientImpl =
+        new XdsClientImpl(
+            "",
+            serverList,
+            XdsClient.XdsChannelFactory.getInstance(),
+            node,
+            syncContext,
+            timeService,
+            new ExponentialBackoffPolicy.Provider(),
+            GrpcUtil.STOPWATCH_SUPPLIER);
+    return new XdsClientWrapperForServerSds(port, xdsClientImpl, timeService);
+  }
+
+  @VisibleForTesting
+  XdsClientWrapperForServerSds(int port, XdsClient xdsClient,
+      ScheduledExecutorService timeService) {
+    this.port = port;
+    this.xdsClient = xdsClient;
+    this.timeService = timeService;
+    xdsClient.watchListenerData(
+        port,
+        new XdsClient.ListenerWatcher() {
+          @Override
+          public void onListenerChanged(XdsClient.ListenerUpdate update) {
+            logger.log(
+                Level.INFO,
+                "Setting myListener from ConfigUpdate listener :{0}",
+                update.getListener().toString());
+            curListener = update.getListener();
+          }
+
+          @Override
+          public void onError(Status error) {
+            // In order to distinguish between IO error and resource not found, set curListener
+            // to null in case of NOT_FOUND
+            if (error.getCode().equals(Status.Code.NOT_FOUND)) {
+              curListener = null;
+            }
+            // TODO(sanjaypujare): Implement logic for other cases based on final design.
+            logger.log(Level.SEVERE, "ListenerWatcher in XdsClientWrapperForServerSds:{0}", error);
+          }
+        });
+  }
+
+  /**
+   * Locates the best matching FilterChain to the channel from the current listener and if found
+   * returns the DownstreamTlsContext from that FilterChain, else null.
+   */
+  @Nullable
+  public DownstreamTlsContext getDownstreamTlsContext(Channel channel) {
```

Comment:
A better API would be to take in an `InetSocketAddress` as the argument instead of a `Channel`, which makes the method specification stronger (i.e., weaker precondition, the method requires less information).

Reply:
You probably missed the comment exchange where @creamsoup talked about this for internal functions.

I do agree that a type like `InetSocketAddress` is better than `Channel` in an API but the filters to be matched have various things like remoteAddress, protocol-type (not just TCP sockets) even if we haven't implemented all of the matching for those fields. And this an internal class (API) and is used by `SdsProtocolNegotiators` which is Netty specific so using `Channel` seems to be the better option.

Reply:
Alright, that't fine. I was just based on what I saw in this method.